### ISSUE TITLE
Map VM errors to `FuelError` error codes

### ISSUE THREAD

**Issue by nedsalk** [FuelLabs/fuels-ts#2467]
- Create `ErrorCode` entry for every VM error
- Every `ErrorCode` entry that maps to a VM error must have a test associated with it:
    - e.g. bring the VM to throw `not enough coins to fit the target` and:
        - Verify that a `FuelError` with the `not-enough-coins` code is thrown
        - Verify that `FuelError.rawError` corresponds to the serialized VM error - this would be done with a `fetch` spy

cc @LuizAsFight 

> [!NOTE]
> Should be done after:
> - https://github.com/FuelLabs/fuels-ts/issues/2521

**Comment by LuizAsFight** [comment]
hey @nedsalk thanks for the heads up

just for context check [this file](https://github.com/FuelLabs/fuels-wallet/blob/4a20b3d8684a07e40bd2f7559032941b5adbe022/packages/app/src/systems/Transaction/services/transaction.tsx#L176) from wallet code

graphql api can return multiple errors, that's why we initially had only `e.response?.errors` , and after the `FuelError` was included we needed to add this conditional, forcing the creation of an array containing the error

do you think it makes sense to create 1 FuelError for each occurrence of errors that came from graphql api? that way returning always an array

**Comment by nedsalk** [comment]
I asked @FuelLabs/client the following question:
> The graphql API returns an errors array when an error happens on the VM. In practice, are there cases where the VM will return more than one error? If there are none, could we make an assumption that there won't be any in the future? Having this assumption would improve ergonomics for users on our side - a thrown error would always be the error, not a wrapper for multiple errors in some cases and the error in others.

And @Dentosal responded:
>  My understanding is that the errors array is part of how GraphQL itself works. If you only request one VM operation, then there can only be one VM error.

@LuizAsFight based on this, we can extract the single error from the graphql `errors` array and throw it. We don't need to handle the _multiple errors_ case because it never happens in reality. The [wallet code](https://github.com/FuelLabs/fuels-wallet/blob/4a20b3d8684a07e40bd2f7559032941b5adbe022/packages/app/src/systems/Transaction/services/transaction.tsx#L176) that you linked to could then maybe be simplified because you wouldn't be working with arrays at all.

**Comment by arboleya** [comment]
I reduced the scope of this issue and moved a portion of it to the below issue, as it will be easier and provide [immediate value](https://github.com/FuelLabs/fuels-ts/issues/2522) even before we can map "all" errors.

 - https://github.com/FuelLabs/fuels-ts/issues/2521

Speaking of _all errors_, do we have a unified list of possible errors the VM can throw, or how else should we know when we mapped _them all_? I suspect this may be something we'll keep mapping as we go.

In this sense, I created a first issue here for a recurring error:
 - https://github.com/FuelLabs/fuels-ts/issues/2522

**Comment by nedsalk** [comment]
> Speaking of all errors, do we have a unified list of possible errors the VM can throw, or how else should we know when we mapped them all? I suspect this may be something we'll keep mapping as we go.

We have a [list](https://github.com/FuelLabs/fuels-ts/blob/master/packages/transactions/src/configs.ts#L46) of VM error codes in our repo already - it's used in a similar error throwing vein, and the full list can be found [here](https://docs.rs/fuel-asm/latest/fuel_asm/enum.PanicReason.html).

**Comment by LuizAsFight** [comment]
@nedsalk one incorrect query can return multiple errors

```
query {
  blocks (last: 10, first: "asd") {
    nodes {
      height
      asds
    }
  }
}
```

response
```
{
  "data": null,
  "errors": [
    {
      "message": "Invalid value for argument \"first\", expected type \"Int\"",
      "locations": [
        {
          "line": 2,
          "column": 20
        }
      ]
    },
    {
      "message": "Unknown field \"asds\" on type \"Block\".",
      "locations": [
        {
          "line": 5,
          "column": 7
        }
      ]
    }
  ]
}
```

**Comment by LuizAsFight** [comment]
I remember to have seen multiple errors before in the wallet also in other scenarios than a wrong query

**Comment by nedsalk** [comment]
Those errors look like graphql-related errors thrown when writing custom queries. The SDK itself internally should never encounter them because we use `Provider.operations`, which is type safe because it's generated via `graphql-codegen`. 

Given that custom queries are out of the SDK's scope, `FuelError` shouldn't be concerned with their failure modes; and even if they were in our scope, it's questionable if they should be integrated into `FuelError` or possibly be a different error class, e.g. `GraphqlError`, because they're not really domain-specific so that they're included into `FuelError` but rather they're related to the GraphQL protocol.

**Comment by LuizAsFight** [comment]
as the graphql natively expose errors array I would just safely keep the same pattern instead of hunt situations that multiple errors can happen. then if multiple errors happen, you're already prepared

but if you wanna do it assuming it will be always only one error and consider the first one, sounds good also, up to you guys

**Comment by Torres-ssf** [comment]
@nedsalk This seems to be a duplicate of https://github.com/FuelLabs/fuels-ts/issues/2208

**Comment by petertonysmith94** [comment]
Going to unassign myself from this for now.
